### PR TITLE
add address sanitizer

### LINKED ISSUE #2142: arena_t testing
## Description

With `reverse_pass_callback` it is easy to write code that passes unit tests but leaks.

We should be able to turn on `-fsanitize=address` to check this (https://releases.llvm.org/6.0.1/tools/clang/docs/AddressSanitizer.html#memory-leak-detection). The address sanitizer page says expect a 2x slowdown. I think we only need to run these on the unit tests, not the probability tests.

I tested this by introducing a bug in [softmax](https://github.com/stan-dev/math/blob/develop/stan/math/rev/fun/softmax.hpp).

If instead of the line:
```
arena_t<Eigen::VectorXd> res_val = softmax(value_of(alpha_arena));
```

We use the line:
```
auto res_val = softmax(value_of(alpha_arena));
```

The code will run and tests pass but there are memory leaks. With the address sanitizer on the tests pass if it is coded correctly and fail with the buggy implementation.

The output looks like:
```
[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (40 ms total)
[  PASSED  ] 1 test.

=================================================================
==4339==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 648 byte(s) in 27 object(s) allocated from:
    #0 0x4e21b0  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x4e21b0)
    #1 0x540143  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x540143)
    #2 0x5b3a15  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x5b3a15)
    #3 0x526278  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x526278)
    #4 0x525f08  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x525f08)
    #5 0x529a53  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x529a53)
    #6 0x52bd52  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x52bd52)
    #7 0x52800b  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x52800b)
    #8 0x522c74  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x522c74)
    #9 0x5227d4  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x5227d4)
    #10 0x521d14  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x521d14)
    #11 0x616911  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x616911)
    #12 0x5e7f6d  (/home/bbales2/math-benchmarks/math/test/unit/math/mix/fun/softmax_test+0x5e7f6d)
```

I don't know if there an equivalent flag in gcc, or if it's even necessary (as long as some of our unit tests build with clang and run with this).

#### Current Version:
v3.3.0

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

Fixes #2142 by adding `-fsanitize=address` to the Full Unit Tests on Jenkins.

This PR will also include fixes for any currently failing tests.

## Tests

/

## Side Effects

/

## Release notes

Added address sanitizer to continuous integration.

## Checklist

- [x] Copyright holder: Rok Češnovar, Univ. of Ljubljana

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/rev/fun/matrix_power.hpp [hunk 1]
```diff
@@ -59,7 +59,7 @@ class matrix_product_vari_n1 : public vari {
 };
 
 template <int R, int C>
-class matrix_product_vari : public vari {
+class matrix_product_vari : public vari, public chainable_alloc {
```
Comment (bbbales2) [inline stan/math/rev/fun/matrix_power.hpp:62]: This allocates a `std::vector` of things in it, so it needs to be on the chainable stack to get cleaned up.

File: stan/math/rev/functor/idas_system.hpp [hunk 2]
```diff
@@ -151,27 +151,36 @@ class idas_system {
         id_(N_VNew_Serial(N_)),
```
Comment (bbbales2) [inline stan/math/rev/functor/idas_system.hpp:151]: All these `N_VNew...` sorta things are basically mallocs, so we gotta free.

File: test/unit/math/rev/core/thread_stack_instance_test.cpp [hunk 3]
```diff
@@ -72,6 +72,7 @@ TEST(thread_stack_instance, repeated_initialize) {
                 ==
 #endif
                 main_ad_stack);
+    delete thread_instance_ptr;
```
Comment (bbbales2) [inline test/unit/math/rev/core/thread_stack_instance_test.cpp:75]: Just looked like an extra pointer that never got free'd.

**Comment by rok-cesnovar** [comment]
@bbbales2 I think the problem is that i put the line in the script parts. Can you move it before the script {} parts? No need for mkdir then.

Else I can fix this in the morning.

**Review by bbbales2 — COMMENTED** [review]
Most of these are just adding `stan::math::recover_memory()` calls to the end of the tests. This cleans up any of the chainable_alloc varis ([here](https://github.com/stan-dev/math/blob/develop/stan/math/rev/core/recover_memory.hpp#L26)). [LDLT_alloc](https://github.com/stan-dev/math/blob/develop/stan/math/rev/fun/LDLT_alloc.hpp) gets used a lot in these tests.

We should probably clean up all the stacks in the autodiff stack destructor: https://github.com/stan-dev/math/blob/develop/stan/math/rev/core/autodiffstackstorage.hpp#L95. I don't think it needs to be this pull.

**Comment by bbbales2** [comment]
@rok-cesnovar doneso

**Comment by rok-cesnovar** [comment]
This seems to pass the sanitizer. I double checked and its being used.

@bbbales2 does this match the results on your machine.

If so, this is ready for review.

**Comment by bbbales2** [comment]
@rok-cesnovar Well feel free to review my changes. I approve of your Jenkins-ing.

**Comment by rok-cesnovar** [comment]
This all looks good. But since its "my" PR you will have to approve, even though my changes in these are minor compared to yours.

Nice that we finally have all the necessary recover memory calls. Even better that we have a way of catching them going forward.

**Comment by bbbales2** [comment]
> This all looks good

Okeedokee. I'll approve and merge when the tests formally pass.

**Comment by bbbales2** [comment]
@serban-nicusor-toptal does uploading the doxygen usually take this long? Or has something gone wrong here?

**Comment by rok-cesnovar** [comment]
There is no free linux agent that does the Doxygen upload. gelman-group-linux seems to have went down a few hours ago and the two AWS instances are running the 6 hours develop distribution tests. They should be done soon.

**Comment by serban-nicusor-toptal** [comment]
The aws instances usually scale a bit slow, I've added now two extra instances so the queue should clear soon.

**Comment by rok-cesnovar** [comment]
Good to go now. Thanks Nic!

**Review by bbbales2 — APPROVED** [review]
Good! (Also @rok-cesnovar says good: https://github.com/stan-dev/math/pull/2146#issuecomment-706563447)

**Pull request merged**